From 149feffd022931e7a16dd17fff33abd4c25e792d Mon Sep 17 00:00:00 2001
From: Hans Pabst <hans.pabst@intel.com>
Date: Tue, 30 Jul 2019 16:35:34 +0200
Subject: [PATCH] Issue #340: fix an issue where realloc return a different
 pointer and subsequently requires to (manually) preserve the original buffer
 content.

---
 src/libxsmm_malloc.c | 7 ++++++-
 version.txt          | 2 +-
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/src/libxsmm_malloc.c b/src/libxsmm_malloc.c
index 27461ec52..f2bcb611f 100644
--- a/src/libxsmm_malloc.c
+++ b/src/libxsmm_malloc.c
@@ -1112,7 +1112,12 @@ LIBXSMM_API_INLINE void* internal_xmalloc_plain(
 #else
     result = realloc((*info)->pointer, size);
     if (NULL != result) {
-      *info = NULL; /* signal no-copy */
+      if (result == (*info)->pointer) {
+        *info = NULL; /* signal no-copy */
+      }
+      else {
+        (*info)->free.function = NULL;
+      }
     }
     else /* error */
 #endif
